consensus/istanbul/backend/engine.go

```diff
@@ -95,15 +96,68 @@ var (
 	nonceAuthVote = hexutil.MustDecode("0xffffffffffffffff") // Magic nonce number to vote on adding a new validator
 	nonceDropVote = hexutil.MustDecode("0x0000000000000000") // Magic nonce number to vote on removing a validator.
 
-	inmemoryAddresses  = 20 // Number of recent addresses from ecrecover
-	recentAddresses, _ = lru.NewARC(inmemoryAddresses)
+	inmemoryBlocks             = 100 // Number of blocks to precompute validators' addresses
+	inmemoryValidatorsPerBlock = 30  // Approximate number of validators' addresses from ecrecover
+	recentAddresses, _         = lru.NewARC(inmemoryBlocks)
+	signatureAddresses, _      = lru.New(inmemoryBlocks * inmemoryValidatorsPerBlock)
 )
 
 // Author retrieves the Klaytn address of the account that minted the given block.
 func (sb *backend) Author(header *types.Header) (common.Address, error) {
 	return ecrecover(header)
 }
 
+// CanVerifyHeadersConcurrently returns true if concurrent header verification possible, otherwise returns false.
+func (sb *backend) CanVerifyHeadersConcurrently() bool {
+	return false
+}
+
+// PreprocessHeaderVerification prepares header verification for heavy computation before synchronous header verification such as ecrecover.
+func (sb *backend) PreprocessHeaderVerification(headers []*types.Header) (chan<- struct{}, <-chan error) {
+	abort := make(chan struct{})
+	results := make(chan error, inmemoryBlocks)
+	go func() {
+		for _, header := range headers {
+			err := sb.computeSignatureAddrs(header)
+
+			select {
+			case <-abort:
+				return
+			case results <- err:
+			}
+		}
+	}()
+	return abort, results
+}
+
+// computeSignatureAddrs computes the addresses of signer and validators and caches them.
+func (sb *backend) computeSignatureAddrs(header *types.Header) error {
+	_, err := ecrecover(header)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the signature from the header extra-data
+	istanbulExtra, err := types.ExtractIstanbulExtra(header)
+	if err != nil {
+		return err
+	}
+
+	proposalSeal := istanbulCore.PrepareCommittedSeal(header.Hash())
+	for _, seal := range istanbulExtra.CommittedSeal {
+		sealStr := hex.EncodeToString(seal)
+		if _, ok := signatureAddresses.Get(sealStr); ok {
```

Comment:
How about using seal as a key for `signatureAddresses`?
```suggestion
		if _, ok := signatureAddresses.Get(seal); ok {
```

Reply:
@KimKyungup The byte slice is not hashable, so the raw seal cannot be used. Any other suggestion?

Reply:
@jeongkyun-oh 
It seems that the key can be byte slice like below? Isn't it?
`hash` is byte array.
`recentAddresses.Add(hash, addr)`